Add goreleaser

suskin opened the pull request:
## Overview

Now that we have our first go command (/ht @turkenh), we want the release process to be as painless and automated as possible. This changeset adds some configuration to do that by integrating with [goreleaser](https://goreleaser.com). Over time, we may make some more customizations and improvements, and add support for more platforms.

The biggest change for this integration is that everything will now be bundled into a single `tar.gz` file, including the go binaries and the bash scripts.

As part of cutting a release, we will want to also update the documentation and the `bootstrap.sh` to default to the latest explicit release. Binary commands will not be available from master yet, though it looks like we can also use goreleaser to build and upload snapshots.

## Testing done

* I followed the release steps locally, and it created a draft release for me in my fork.
* I created the tar.gz files a few times and tried them with the `tar` command from `bootstrap.sh` (only on darwin; if someone wants to test on linux I would appreciate that)

suskin committed:
release: add goreleaser config file

We want to try to automate the build and relase of multiple built
binaries for our go-based tools. This change adds a first pass at a
configuration for the goreleaser tool, which should make that more
convenient.

Signed-off-by: suskin <813571+suskin@users.noreply.github.com>

suskin committed:
release: add configuration for building and bundling

Goreleaser has some configuration directives for building separate
commands, and bundling archives. This adds some configuration for what
we're trying to do, which is build trace from a specific subcommand
directory, and bundle it in a shared tarball.

Signed-off-by: suskin <813571+suskin@users.noreply.github.com>

suskin committed:
release: add configuration for building and downloading

Now that we have goreleaser working in general, this change adds some
refinements so namings, folders, and stuff all work properly.

Signed-off-by: suskin <813571+suskin@users.noreply.github.com>

suskin committed:
release: bundle bash scripts in the archive

goreleaser supports bundling other files in the archive alongside go
binaries, so this change takes advantage of that to bundle all of the
crossplane-cli commands into the cheesy tarball.

Signed-off-by: suskin <813571+suskin@users.noreply.github.com>

suskin committed:
release: add section and comments for release

Now that the binary seems to be in good shape, we also want to have
configuration for the release on github.

Signed-off-by: suskin <813571+suskin@users.noreply.github.com>

turkenh reviewed (approved):
LGTM!
Also tested for linux.

Thanks!

displague commented:
https://github.com/crossplaneio/crossplane-cli/pull/17/files#diff-3af436e6760a0c5f1e3d16a3384853a8L22 should be a for loop over the binary names.. Just an observation on the existing code, not a blocker for the PR.

Bootstrap.sh should also start with a comment describing its function.

jbw976 commented on go.mod:
```diff
@@ -21,7 +21,6 @@ require (
 	golang.org/x/crypto v0.0.0-20191011191535-87dc89f01550 // indirect
 	golang.org/x/net v0.0.0-20191014212845-da9a3fd4c582 // indirect
 	golang.org/x/time v0.0.0-20190921001708-c4c64cad1fd0 // indirect
-	gopkg.in/yaml.v2 v2.2.4
```
why did go.mod and go.sum change when there don't appear to be any source code changes?

suskin replied:
Running the release runs `go mod tidy`, and that made these changes

jbw976 commented on .goreleaser.yml:
```diff
@@ -0,0 +1,111 @@
+# For more about how to configure goreleaser, see:
+# https://goreleaser.com/customization/
+#
+# To release:
```
do you think we'll plan on creating release branches as well, or just do all releases from tags on commits in master?

suskin replied:
We probably don't need the release branches anymore, but we might still create them because they're useful if we ever want to make a patch for an old release while we're working on a new release.

jbw976 reviewed (commented):
Looks pretty reasonable @suskin.  Do you have an example of what a published release looks like?  I didn't see any on your fork https://github.com/suskin/crossplane-cli

suskin reviewed (commented):
> Do you have an example of what a published release looks like? I didn't see any on your fork https://github.com/suskin/crossplane-cli

I had one in a draft that wasn't published, so here's the published version: https://github.com/suskin/crossplane-cli/releases/tag/v0.2.0-example /cc @jbw976

suskin committed:
release: exclude pull request merges from the changelog

Previously, commits which merged pull requests were showing up in the
changelog. Now they won't.

Signed-off-by: suskin <813571+suskin@users.noreply.github.com>

suskin commented:
> https://github.com/crossplaneio/crossplane-cli/pull/17/files#diff-3af436e6760a0c5f1e3d16a3384853a8L22 should be a for loop over the binary names.. Just an observation on the existing code, not a blocker for the PR.
> 
> Bootstrap.sh should also start with a comment describing its function.

Thanks for calling these out - I'll make a note of them so we can address them in a future changeset!
